test/int/wizards/breakpoints/implementation/internalFileBreakpointsWizard.ts

```diff
@@ -0,0 +1,84 @@
+    assertIsHitThenResumeWhen(breakpoint: BreakpointWizard, lastActionToMakeBreakpointHit: () => Promise<void>, expectedStackTrace: string): Promise<void>;
+
+    onBreakpointStatusChange(breakpointStatusChanged: DebugProtocol.BreakpointEvent): void;
+}
+
+export type CurrentBreakpointsMapping = ValidatedMap<BreakpointWizard, DebugProtocol.Breakpoint>;
+
+export type StateChanger = (newState: IPerformChangesImmediatelyOrBatchState) => void;
+
+export class InternalFileBreakpointsWizard {
+    private _state: IPerformChangesImmediatelyOrBatchState = new PerformChangesImmediatelyState(this._breakpointsWizard, this, new ValidatedMap());
+
+    public constructor(private readonly _breakpointsWizard: BreakpointsWizard, public readonly client: ExtendedDebugClient, public readonly filePath: string) { }
+
+    public async hitCountBreakpoint(options: { lineText: string; hitCountCondition: string; name: string }): Promise<BreakpointWizard> {
+        const position = await findPositionOfTextInFile(this.filePath, options.lineText);
+        return new BreakpointWizard(this, position, new PauseOnHitCount(options.hitCountCondition), options.name);
+    }
+
+    public async set(breakpointWizard: BreakpointWizard): Promise<void> {
+        await this._state.set(breakpointWizard);
+    }
+
+    public async unset(breakpointWizard: BreakpointWizard): Promise<void> {
+        await this._state.unset(breakpointWizard);
+    }
+
+    public async waitUntilVerified(breakpoint: BreakpointWizard): Promise<void> {
+        await this._state.waitUntilVerified(breakpoint);
+    }
+
+    public assertIsVerified(breakpoint: BreakpointWizard): void {
+        this._state.assertIsVerified(breakpoint);
+    }
+
+    public async assertIsHitThenResumeWhen(breakpoint: BreakpointWizard, lastActionToMakeBreakpointHit: () => Promise<void>, expectedStackTrace: string): Promise<void> {
+        return this._state.assertIsHitThenResumeWhen(breakpoint, lastActionToMakeBreakpointHit, expectedStackTrace);
+    }
+
+    public onBreakpointStatusChange(breakpointStatusChanged: DebugProtocol.BreakpointEvent): void {
+        this._state.onBreakpointStatusChange(breakpointStatusChanged);
+    }
+
+    public async batch<T>(batchAction: (fileBreakpointsWizard: FileBreakpointsWizard) => PromiseOrNot<T>): Promise<T> {
+        const batchingUpdates = new BatchingUpdatesState(this, this._state.currentBreakpointsMapping);
+        this._state = batchingUpdates;
+        const result = await batchAction(new FileBreakpointsWizard(this));
+        await batchingUpdates.processBatch(); // processBatch calls sendBreakpointsToClient which will change the state back to PerformChangesImmediatelyState
+        return result;
+    }
+
+    public async sendBreakpointsToClient(update: BreakpointsUpdate): Promise<void> {
+        return new BreakpointsUpdater(this._breakpointsWizard, this, this.client, update, state => this._state = state).update();
```

Comment:
Do we need to create a new BreakpointsUpdater on every call?

Reply:
Could you please elaborate on what advantages you see if we change this to use the same instance on every call?

Reply:
Code clarity, mainly. Something like:

`this.breakpointUpdater.send(newBreakpoints)`

is much easier to understand at a quick glance than the complicated constructor->update().

Reply:
Changed